getting-started/sessions_new/main.go

```diff
@@ -0,0 +1,203 @@
+
+	projectID := os.Getenv("GOOGLE_CLOUD_PROJECT")
+	if projectID == "" {
+		log.Fatal("GOOGLE_CLOUD_PROJECT must be set")
+	}
+
+	// collectionID is a non-empty identifier for this app, it is used as the Firestore
+	// collection name that stores the sessions.
+	//
+	// Set it to something more descriptive for your app.
+	collectionID := "hello-views"
+
+	a, err := newApp(projectID, collectionID)
+	if err != nil {
+		log.Fatalf("newApp: %v", err)
+	}
+
+	http.HandleFunc("/", a.index)
+
+	log.Printf("Listening on port %s", port)
+	if err := http.ListenAndServe(":"+port, nil); err != nil {
+		log.Fatal(err)
+	}
+}
+
+// newApp creates a new app.
+func newApp(projectID, collectionID string) (*app, error) {
+	ctx := context.Background()
+	client, err := firestore.NewClient(ctx, projectID)
+	if err != nil {
+		log.Fatalf("firestore.NewClient: %v", err)
+	}
+
+	tmpl, err := template.New("Index").Parse(`<body>{{.Views}} {{if eq .Views 1}}view{{else}}views{{end}} for "{{.Greetings}}"</body>`)
+	if err != nil {
+		log.Fatalf("template.New: %v", err)
+	}
+
+	return &app{
+		client:       client,
+		tmpl:         tmpl,
+		collectionID: collectionID,
+	}, nil
+}
+
+// [END getting_started_sessions_main]
+
+// [START getting_started_sessions_handler]
+
+// index uses sessions to assign users a random greeting and keep track of
+// views.
+func (a *app) index(w http.ResponseWriter, r *http.Request) {
+	if r.RequestURI != "/" {
+		return
+	}
+
+	ctx := context.Background()
+
+	var id string
+	var session session
+	var cookie *http.Cookie
+	var isNewSession bool
+
+	// cookieName is a non-empty identifier for this app, it is used as the key name
+	// that contains the session's id value.
+	//
+	// Set it to something more descriptive for your app.
+	cookieName := "session_id"
+
+	// If err is different to nil, it means the cookie has not been set, so it will be created.
+	cookie, err := r.Cookie(cookieName)
+	if err != nil {
+		if errors.Is(err, http.ErrNoCookie) {
+			// isNewSession flag is set to true
+			isNewSession = true
+		} else {
+			log.Printf("Error getting cookie: %v", err)
+			http.Error(w, "Error accessing session", http.StatusInternalServerError)
+			return
+		}
+	}
+
+	// If isNewSession flag is true, the session will be created
+	if isNewSession {
+		// Get unique id for new document
+		id = a.client.Collection(a.collectionID).NewDoc().ID
+
+		session.Greetings = greetings[rand.Intn(len(greetings))]
+		session.Views = 1
+
+		// New document is created
+		_, err := a.client.Collection(a.collectionID).Doc(id).Set(ctx, session)
+		if err != nil {
+			log.Printf("client.Collection.Doc.Set error: %v", err)
+			http.Error(w, "Error creating session", http.StatusInternalServerError)
+			return
+		}
+
+		// Cookie is set
+		cookie = &http.Cookie{
+			Name:  cookieName,
+			Value: id,
+		}
+		http.SetCookie(w, cookie)
+	} else {
+		// The session exists
+
+		// Get session
+		doc, err := a.client.Collection(a.collectionID).Doc(cookie.Value).Get(ctx)
```

Comment:
issue: same pattern as if block: create a new client, be sure to defer Close(), hold onto the DocRef.